drop-in/src/main/java/com/adyen/checkout/dropin/service/DropInService.kt

```diff
@@ -31,151 +32,56 @@ import org.json.JSONObject
-                if (paymentComponentDataForRequest == null) {
-                    handleDropInServiceResult(
-                        DropInServiceResult.Error(
-                            reason = "DropInService Error. No content in PAYMENT_COMPONENT_DATA_EXTRA_KEY"
-                        )
-                    )
-                } else {
-                    askPaymentsCall(paymentComponentDataForRequest)
-                }
-            }
-            DETAILS_REQUEST -> {
-                val detailsString = intent.getStringExtra(DETAILS_EXTRA_KEY) ?: ""
-                val details = JSONObject(detailsString)
-                askDetailsCall(details)
-            }
-        }
+    override fun onCreate() {
+        Logger.d(TAG, "onCreate")
+        super.onCreate()
     }
 
     override fun onDestroy() {
-        super.onDestroy()
-        // Android will only call new tasks from enqueueWork if the previous one has finished nas the service is destroyed
         Logger.d(TAG, "onDestroy")
+        super.onDestroy()
     }
 
-    /**
-     * Call this method for asynchronous handling of [makePaymentsCall]
-     */
-    @Suppress("MemberVisibilityCanBePrivate")
-    protected fun asyncCallback(dropInServiceResult: DropInServiceResult) {
-        handleDropInServiceResult(dropInServiceResult)
-    }
-
-    private fun askPaymentsCall(paymentComponentData: PaymentComponentData<in PaymentMethodDetails>) {
+    internal fun requestPaymentsCall(paymentComponentData: PaymentComponentData<*>) {
         Logger.d(TAG, "askPaymentsCall")
 
         // Merchant makes network call
-        val paymentsDropInServiceResult = makePaymentsCall(PaymentComponentData.SERIALIZER.serialize(paymentComponentData))
-
-        handleDropInServiceResult(paymentsDropInServiceResult)
+        makePaymentsCall(PaymentComponentData.SERIALIZER.serialize(paymentComponentData))
     }
 
-    private fun askDetailsCall(details: JSONObject) {
+    internal fun requestDetailsCall(details: JSONObject) {
         Logger.d(TAG, "askDetailsCall")
 
         // Merchant makes network call
-        val detailsDropInServiceResult = makeDetailsCall(details)
-
-        handleDropInServiceResult(detailsDropInServiceResult)
+        makeDetailsCall(details)
     }
 
-    private fun handleDropInServiceResult(dropInServiceResult: DropInServiceResult?) {
-        if (dropInServiceResult == null) {
-            // Make sure people don't return Null from Java code
-            throw CheckoutException("DropInServiceResult result from DropInService cannot be null.")
-        }
-        Logger.d(TAG, "handleDropInServiceResult - ${dropInServiceResult::class.simpleName}")
-
-        // if type is WAIT do nothing and wait for async callback.
-        if (dropInServiceResult !is DropInServiceResult.Wait) {
+    protected fun sendResult(result: DropInServiceResult) {
+        ThreadManager.MAIN_HANDLER.post {
```

Comment:
@caiofaustino the `DropInActivity` should handle the result on the main thread to be able to send an intent and finish and so on. So I wondered where we should do the switch:
* If we leave it up to the merchant to switch back to the main thread, we need to add a check and throw an exception
* We could do the switch in the `DropInActivity` since it is the component that needs to handle the result on the main thread.
* But by doing the switch in the `DropInService` we make sure the `DropInActivity` does not handle any threading logic since it's a task that seems more suitable with a `Service` than an `Activity`.
* Also am I doing the switch the right way or do we have a better way?

WDYT?

Reply:
All right so just to summarize the discussion so we don't forget. Something like this:
```kotlin
internal open fun requestPaymentsCall(
    paymentComponentState: paymentComponentState,
    paymentDetails: JSONObject
) {
    ThreadManager.EXECUTOR.submit {
        val callResult = makeDetailsCall(details)
        handleResult()
    }
}
```

1. We will keep the current approach of having the default method being called on a background thread.
2. They can override `requestPaymentsCall` to handle the call asynchronously, and then they can manually call `handleResult()
3. `makeDetailsCall` default implementation throws an exception.
4. The DropInActivity can observe a `resultLiveData` on the Service. `handleResult()` will post to the livedata.
5. Calling `handleResult()` with an Error can prompt the Drop-in to terminate.